mono/metadata/icall.c

```diff
@@ -6891,53 +6891,83 @@ ves_icall_System_Environment_GetCommandLineArgs (MonoError *error)
 	return result;
 }
 
-#ifndef HOST_WIN32
-static MonoArray *
-mono_icall_get_environment_variable_names (MonoError *error)
-{
-	MonoArray *names;
-	MonoDomain *domain;
-	MonoString *str;
-	gchar **e, **parts;
-	int n;
+MonoArrayHandle
+ves_icall_System_Environment_GetEnvironmentVariableNames (MonoError *error)
+{
+	MonoDomain * const domain = mono_domain_get ();
+	gsize n = 0;
+	MonoStringHandle str = MONO_HANDLE_NEW (MonoString, NULL);
+
+#ifdef HOST_WIN32
+	WCHAR * const env_strings = GetEnvironmentStringsW ();
+
+	if (env_strings) {
+		WCHAR const *env_string = env_strings;
+		while (*env_string) {
+			/* weird case that MS seems to skip (per drive letter current working directory) */
+			if (*env_string != '=')
+				n++;
+			while (*env_string)
+				env_string++;
+			env_string++;
+		}
+	}
 
-	error_init (error);
-	n = 0;
-	for (e = environ; *e != 0; ++ e)
-		++ n;
+	MonoArrayHandle names = mono_array_new_handle (domain, mono_defaults.string_class, n, error);
+	return_val_if_nok (error, NULL_HANDLE_ARRAY);
 
-	domain = mono_domain_get ();
-	names = mono_array_new_checked (domain, mono_defaults.string_class, n, error);
-	return_val_if_nok (error, NULL);
+	if (env_strings) {
+		n = 0;
+		WCHAR const *env_string = env_strings;
+		while (*env_string) {
+			/* weird case that MS seems to skip (per drive letter current working directory) */
+			if (*env_string != '=') {
+				WCHAR const * const equal_str = wcschr (env_string, '=');
+				g_assert (equal_str);
+				mono_string_new_utf16_to_handle (str, domain, env_string, (gsize)(equal_str - env_string), error);
+				goto_if_nok (error, exit);
+				MONO_HANDLE_ARRAY_SETREF (names, n, str);
+				n++;
+			}
+			while (*env_string)
+				env_string++;
+			env_string++;
+		}
+
+	}
+
+exit:
+	if (env_strings)
+		FreeEnvironmentStringsW (env_strings);
+	return is_ok (error) ? names : NULL_HANDLE_ARRAY;
+#else
+	char **e;
+
+	for (e = environ; *e; ++e)
+		++n;
+
+	MonoArrayHandle names = mono_array_new_handle (domain, mono_defaults.string_class, n, error);
+	return_val_if_nok (error, NULL_HANDLE_ARRAY);
 
 	n = 0;
 	for (e = environ; *e != 0; ++ e) {
-		parts = g_strsplit (*e, "=", 2);
-		if (*parts != 0) {
-			str = mono_string_new_checked (domain, *parts, error);
+		char **parts = g_strsplit (*e, "=", 2);
+		if (*parts) {
+			mono_string_new_utf8_to_handle (str, domain, *parts, strlen (*parts), error);
```

Comment:
Same as https://github.com/mono/mono/pull/11407/files#diff-eba0fd0052bf3ec9dd60f2f3e90f19e1R6927, revert to using `mono_string_new_utf8_handle`.